### PR TITLE
throw exceptions when errors happen

### PR DESCRIPTION
**Pull request by Xiaochao8**

### CODE REVIEW

File: lib/sgtn-client/loader/source_comparer.rb [hunk 1]
```diff
@@ -10,22 +10,34 @@ module TranslationLoader
...
 
         old_source_bundle_thread = Thread.new { load_bundle(component, CONSTS::OLD_SOURCE_LOCALE) }
+        source_bundle_thread = Thread.new { load_bundle(component, LocaleUtil.get_source_locale) }
         translation_bundle = super(component, locale)
-        source_bundle = get_bundle(component, LocaleUtil.get_source_locale) # source is in cache and does not expire.
-        old_source_bundle = old_source_bundle_thread.value
+
+        begin
+          old_source_bundle = old_source_bundle_thread.value
+          source_bundle = source_bundle_thread.value
```
Comment (jessiejuachon) [inline lib/sgtn-client/loader/source_comparer.rb:25]: why another thread again?

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

File: lib/sgtn-client/loader/source_comparer.rb (hunk 1 above)
Comment (Xiaochao8) [reply lib/sgtn-client/loader/source_comparer.rb]: Remote server has source, a seperate thread can speed up.

File: lib/sgtn-client/loader/source.rb [hunk 2]
```diff
@@ -22,11 +22,18 @@ def load_bundle(component, locale = nil)
...
 
         (@source_bundle_path + component).glob('**/*.{yml, yaml}') do |f|
-          bundle = YAML.load(File.read(f))
-          total_messages.merge!(bundle&.first&.last) # TODO: Warning about inconsistent source locale
+          bundle = YAML.safe_load(File.read(f))
+          messages = bundle&.first&.last # TODO: Warn about inconsistent source locale
+          if messages.is_a?(Hash)
+            total_messages.merge!(messages)
+          else
+            SgtnClient.logger.error "[#{method(__callee__).owner}.#{__callee__}] Illegal bundle data in #{f}"
```
Comment (jessiejuachon) [inline lib/sgtn-client/loader/source.rb:30]: Change the word 'illegal' to 'invalid'.

File: lib/sgtn-client/loader/source.rb [hunk 3]
```diff
@@ -22,11 +22,18 @@ def load_bundle(component, locale = nil)
         total_messages = {}
 
         (@source_bundle_path + component).glob('**/*.{yml, yaml}') do |f|
-          bundle = YAML.load(File.read(f))
-          total_messages.merge!(bundle&.first&.last) # TODO: Warning about inconsistent source locale
+          bundle = YAML.safe_load(File.read(f))
+          messages = bundle&.first&.last # TODO: Warn about inconsistent source locale
+          if messages.is_a?(Hash)
+            total_messages.merge!(messages)
+          else
```
Comment (jessiejuachon) [inline lib/sgtn-client/loader/source.rb:29]: When will messages not be a Hash?

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

**Comment by jessiejuachon** [comment]
I tested this in a sample app. translate method always returns the key.

File: lib/sgtn-client/loader/source.rb (hunk 3 above)
Comment (Xiaochao8) [reply lib/sgtn-client/loader/source.rb:29]: Messages will be nil when the source file is empty.

File: lib/sgtn-client/loader/source.rb (hunk 2 above)
Comment (Xiaochao8) [reply lib/sgtn-client/loader/source.rb]: fixed

**Review by jessiejuachon — APPROVED** [review]

**Review by deminyan — APPROVED** [review]

**Pull request merged**